File: frontend/passes/pipeline.go
package passes

import (
	"github.com/mattcarp12/maml/frontend/mir"
	"github.com/mattcarp12/maml/frontend/parser/ast"
	"github.com/mattcarp12/maml/frontend/types"
)

type PassConfig struct {
	Liveness    bool
	Borrow      bool
	LinearLower bool
	DropInject  bool
}

// DefaultConfig returns the production configuration, now ARC-free.
func DefaultConfig() PassConfig {
	return PassConfig{
		Liveness:    true,
		Borrow:      true,
		LinearLower: true,
		DropInject:  true,
	}
}

func RunPasses(g *mir.Graph, locals map[string]types.Type, cfg PassConfig) []ast.CompileError {
	if g == nil {
		return nil
	}

	var livenessRes *LivenessResult
	if cfg.Liveness {
		livenessRes = AnalyzeLiveness(g, locals)
	}

	var borrowErrs []ast.CompileError
	if cfg.Borrow && livenessRes != nil {
		borrowAnalyzer := New()
		borrowErrs = borrowAnalyzer.Analyze(g, locals, livenessRes)
	}

	if cfg.DropInject && livenessRes != nil {
		InjectDrops(g, locals, livenessRes)
	}

	if cfg.LinearLower {
		LowerLinearTypes(g)
	}

	return borrowErrs
}

File: passes/src/pipeline.cpp
#include "pipeline.h"
#include "borrow.h"
#include "drop_injector.h"
#include "linear_lower.h"
#include "liveness.h"

namespace maml::passes {

PassConfig defaultConfig() { return PassConfig { true, true, true, true }; }

std::vector<ast::CompileError> runPasses(mir::Graph* g,
    std::unordered_map<SymID, const types::Type*>& locals, SymbolTable& sym,
    types::TypeRegistry& reg, const PassConfig& cfg)
{
    if (!g)
        return {};

    std::unique_ptr<LivenessResult> livenessRes = nullptr;
    if (cfg.liveness) {
        LivenessAnalyzer livenessAnalyzer(sym);
        livenessRes = std::make_unique<LivenessResult>(livenessAnalyzer.analyzeLiveness(g, locals));
    }

    std::vector<ast::CompileError> borrowErrs;
    if (cfg.borrow && livenessRes) {
        BorrowAnalyzer borrowAnalyzer(sym);
        borrowErrs = borrowAnalyzer.analyze(g, locals, *livenessRes);
    }

    if (cfg.dropInject && livenessRes) {
        DropInjector dropInjector(sym, reg);
        dropInjector.injectDrops(g, locals, *livenessRes);
    }

    if (cfg.linearLower) {
        lowerLinearTypes(g);
    }

    return borrowErrs;
}

} // namespace maml::passes